bug: dbt-postgres utility inherit_from causing env var issues

### Meltano Version

3.7.7

### Python Version

3.12

### Bug scope

Configuration (settings parsing, validation, etc.)

### Operating System

Linux - Ubuntu 20.04

### Description

I was trying to use the dbt-postgres utility in Meltano and leverage inherit_from: in my dbt_project.yml so that I could share common configuration across multiple projects. When I run meltano invoke dbt-postgres run (or call it via a Meltano job), dbt immediately fails with a parsing error about a missing environment variable instead of loading the inherited config. I expected Meltano’s dbt-postgres integration to honor inherit_from.

```
2025-06-01T20:58:31.522938Z [info     ] 20:58:31  Running with dbt=1.9.4 cmd_type=command name=dbt-postgres stdio=stderr
2025-06-01T20:58:31.527514Z [info     ] 20:58:31  Encountered an error: cmd_type=command name=dbt-postgres stdio=stderr
2025-06-01T20:58:31.527760Z [info     ] Parsing Error                  cmd_type=command name=dbt-postgres stdio=stderr
2025-06-01T20:58:31.528531Z [info     ]   Env var required but not provided: 'DBT_POSTGRES_HOST' cmd_type=command name=dbt-postgres stdio=stderr
2025-06-01T20:58:31.638849Z [error    ] Block run completed.           block_type=InvokerCommand err=RunnerError('`dbt-postgres select-ndo` failed with exit code: 2') exit_codes={} set_number=2 success=False
Need help fixing this problem? Visit http://melta.no/ for troubleshooting steps, or to
join our friendly Slack community.

Run invocation could not be completed as block failed: `dbt-postgres select-ndo` failed with exit code: 2
make: *** [pipeline] Error 1
```

### Code

```yaml
utilities:
  - name: dbt-postgres
    variant: dbt-labs
    pip_url: dbt-core dbt-postgres meltano-dbt-ext~=0.3.0
  - name: dbt-postgres--example
    inherit_from: dbt-postgres
```

### Link to Slack/Linen
(https://discuss.meltano.com/t/27037771/hello-i-am-running-into-an-issue-when-trying-to-inherit-the-)
_No response_

--- a/src/meltano/core/plugin/settings_service.py
+++ b/src/meltano/core/plugin/settings_service.py
@@ -49,7 +49,7 @@ class PluginSettingsService(SettingsService):
         if self.inherited_settings_service:
             # Safe shallow copy clone to avoid reference mutability side-effects
             parent_env = dict(self.inherited_settings_service.env_override)
-            
+
         self.env_override = {
             # parent/inherited configuration environment values:
             **parent_env,

--- a/src/meltano/core/plugin/settings_service.py
+++ b/src/meltano/core/plugin/settings_service.py
@@ -47,8 +47,9 @@ class PluginSettingsService(SettingsService):
         # Fetch environment configs from parent plugin if inherit_from is used
         parent_env = {}
         if self.inherited_settings_service:
-            parent_env = self.inherited_settings_service.env_override
-
+            # Safe shallow copy clone to avoid reference mutability side-effects
+            parent_env = dict(self.inherited_settings_service.env_override)
+            
         self.env_override = {
             # parent/inherited configuration environment values:
             **parent_env,

--- a/src/meltano/core/plugin/settings_service.py
+++ b/src/meltano/core/plugin/settings_service.py
@@ -44,7 +44,14 @@ class PluginSettingsService(SettingsService):
                 self.plugin.name,
             )
 
+        # Fetch environment configs from parent plugin if inherit_from is used
+        parent_env = {}
+        if self.inherited_settings_service:
+            parent_env = self.inherited_settings_service.env_override
+
         self.env_override = {
+            # parent/inherited configuration environment values:
+            **parent_env,
             # project level environment variables:
             **self.project.settings.env,
             # project level settings as env vars (e.g. `MELTANO_PROJECT_ID`):

--- a/src/meltano/core/plugin/settings_service.py
+++ b/src/meltano/core/plugin/settings_service.py
@@ -44,15 +44,13 @@ class PluginSettingsService(SettingsService):
                 self.plugin.name,
             )
 
-        # Fetch environment configs from parent plugin if inherit_from is used
-        parent_env = {}
-        if self.inherited_settings_service:
-            # Safe shallow copy clone to avoid reference mutability side-effects
-            parent_env = dict(self.inherited_settings_service.env_override)
-
         self.env_override = {
-            # parent/inherited configuration environment values:
-            **parent_env,
+            # Fetch environment configs from parent plugin if inherit_from is used
+            **(
+                self.inherited_settings_service.env_override
+                if self.inherited_settings_service
+                else {}
+            ),
             # project level environment variables:
             **self.project.settings.env,
             # project level settings as env vars (e.g. `MELTANO_PROJECT_ID`):

--- a/tests/meltano/core/plugin/test_plugin_settings.py
+++ b/tests/meltano/core/plugin/test_plugin_settings.py
@@ -1095,12 +1095,28 @@ class TestPluginSettingsService:
         self,
         plugin_settings_service_factory,
         inherited_tap: ProjectPlugin,
+        monkeypatch: pytest.MonkeyPatch,
     ) -> None:
         """Test child plugins safely clone parent env_override."""
         # 1. Initialize the child service (which builds its parent)
         child_service = plugin_settings_service_factory(inherited_tap)
         parent_service = child_service.inherited_settings_service
+        assert parent_service is not None
 
         # 2. Verify mutation isolation (child state won't poison parent)
         child_service.env_override["CHILD_EXCLUSIVE_KEY"] = "isolated"
         assert "CHILD_EXCLUSIVE_KEY" not in parent_service.env_override
+
+        # 3. Verify parent's `env:` block inheritance
+        parent = inherited_tap.parent
+        monkeypatch.setattr(parent, "env", {"PARENT_ENV_VAR": "from_parent_env"})
+        child_service = plugin_settings_service_factory(inherited_tap)
+        assert child_service.env_override.get("PARENT_ENV_VAR") == "from_parent_env"
+
+        # 4. Child's own env_override kwarg takes priority over parent's `env:` block.
+        monkeypatch.setattr(parent, "env", {"SHARED_ENV_VAR": "from_parent"})
+        child_service = plugin_settings_service_factory(
+            inherited_tap,
+            env_override={"SHARED_ENV_VAR": "from_child"},
+        )
+        assert child_service.env_override.get("SHARED_ENV_VAR") == "from_child"

--- a/tests/meltano/core/plugin/test_plugin_settings.py
+++ b/tests/meltano/core/plugin/test_plugin_settings.py
@@ -1096,14 +1096,14 @@ class TestPluginSettingsService:
         plugin_settings_service_factory,
         inherited_tap: ProjectPlugin,
     ) -> None:
-        """Test that child plugins safely clone parent env_override instead of referencing it."""
-        # 1. Initialize the child service (which automatically builds its parent)
+        """Test child plugins safely clone parent env_override."""
+        # 1. Initialize the child service (which builds its parent)
         child_service = plugin_settings_service_factory(inherited_tap)
         parent_service = child_service.inherited_settings_service
-
-        # 2. Assert they are NOT the exact same dictionary object in memory (shallow copy check)
+        
+        # 2. Assert they are distinct memory objects (shallow copy)
         assert child_service.env_override is not parent_service.env_override
 
-        # 3. Verify mutation isolation (modifying child does not poison parent state)
+        # 3. Verify mutation isolation (child state won't poison parent)
         child_service.env_override["CHILD_EXCLUSIVE_KEY"] = "isolated"
         assert "CHILD_EXCLUSIVE_KEY" not in parent_service.env_override

--- a/tests/meltano/core/plugin/test_plugin_settings.py
+++ b/tests/meltano/core/plugin/test_plugin_settings.py
@@ -1094,22 +1094,16 @@ class TestPluginSettingsService:
     def test_inherited_env_override(
         self,
         plugin_settings_service_factory,
-        tap: ProjectPlugin,
         inherited_tap: ProjectPlugin,
-        monkeypatch,
     ) -> None:
-        """Test that child plugins correctly inherit env_override from their parent."""
-        # 1. Set a specific env var on the parent plugin
-        parent_service = plugin_settings_service_factory(tap)
-        parent_env_key = "DBT_POSTGRES_HOST"
-        parent_env_val = "parent-host"
-
-        # Manually inject into the parent's environment context
-        monkeypatch.setitem(parent_service.env_override, parent_env_key, parent_env_val)
-
-        # 2. Initialize the child service (which inherits from tap)
+        """Test that child plugins safely clone parent env_override instead of referencing it."""
+        # 1. Initialize the child service (which automatically builds its parent)
         child_service = plugin_settings_service_factory(inherited_tap)
-
-        # 3. Assert that the child now sees the parent's environment variable
-        # Our fix ensures the child service merges the parent's env_override
-        assert child_service.env_override.get(parent_env_key) == parent_env_val
+        parent_service = child_service.inherited_settings_service
+        
+        # 2. Assert they are NOT the exact same dictionary object in memory (shallow copy check)
+        assert child_service.env_override is not parent_service.env_override
+        
+        # 3. Verify mutation isolation (modifying child does not poison parent state)
+        child_service.env_override["CHILD_EXCLUSIVE_KEY"] = "isolated"
+        assert "CHILD_EXCLUSIVE_KEY" not in parent_service.env_override

--- a/tests/meltano/core/plugin/test_plugin_settings.py
+++ b/tests/meltano/core/plugin/test_plugin_settings.py
@@ -1090,3 +1090,26 @@ class TestPluginSettingsService:
         )
         subject.set("stacked_env_var", "${NONEXISTENT_ENV_VAR}")
         assert subject.get("stacked_env_var") is None
+
+    def test_inherited_env_override(
+        self,
+        plugin_settings_service_factory,
+        tap: ProjectPlugin,
+        inherited_tap: ProjectPlugin,
+        monkeypatch,
+    ) -> None:
+        """Test that child plugins correctly inherit env_override from their parent."""
+        # 1. Set a specific env var on the parent plugin
+        parent_service = plugin_settings_service_factory(tap)
+        parent_env_key = "DBT_POSTGRES_HOST"
+        parent_env_val = "parent-host"
+        
+        # Manually inject into the parent's environment context
+        monkeypatch.setitem(parent_service.env_override, parent_env_key, parent_env_val)
+
+        # 2. Initialize the child service (which inherits from tap)
+        child_service = plugin_settings_service_factory(inherited_tap)
+
+        # 3. Assert that the child now sees the parent's environment variable
+        # Our fix ensures the child service merges the parent's env_override
+        assert child_service.env_override.get(parent_env_key) == parent_env_val

--- a/tests/meltano/core/plugin/test_plugin_settings.py
+++ b/tests/meltano/core/plugin/test_plugin_settings.py
@@ -1100,10 +1100,10 @@ class TestPluginSettingsService:
         # 1. Initialize the child service (which automatically builds its parent)
         child_service = plugin_settings_service_factory(inherited_tap)
         parent_service = child_service.inherited_settings_service
-        
+
         # 2. Assert they are NOT the exact same dictionary object in memory (shallow copy check)
         assert child_service.env_override is not parent_service.env_override
-        
+
         # 3. Verify mutation isolation (modifying child does not poison parent state)
         child_service.env_override["CHILD_EXCLUSIVE_KEY"] = "isolated"
         assert "CHILD_EXCLUSIVE_KEY" not in parent_service.env_override

--- a/tests/meltano/core/plugin/test_plugin_settings.py
+++ b/tests/meltano/core/plugin/test_plugin_settings.py
@@ -1101,9 +1101,6 @@ class TestPluginSettingsService:
         child_service = plugin_settings_service_factory(inherited_tap)
         parent_service = child_service.inherited_settings_service
 
-        # 2. Assert they are distinct memory objects (shallow copy)
-        assert child_service.env_override is not parent_service.env_override
-
-        # 3. Verify mutation isolation (child state won't poison parent)
+        # 2. Verify mutation isolation (child state won't poison parent)
         child_service.env_override["CHILD_EXCLUSIVE_KEY"] = "isolated"
         assert "CHILD_EXCLUSIVE_KEY" not in parent_service.env_override

--- a/tests/meltano/core/plugin/test_plugin_settings.py
+++ b/tests/meltano/core/plugin/test_plugin_settings.py
@@ -26,7 +26,7 @@ from meltano.core.settings_store import (
 from meltano.core.utils import EnvironmentVariableNotSetError
 
 if t.TYPE_CHECKING:
-    from collections.abc import Callable, Generator
+    from collections.abc import Generator
 
     from sqlalchemy.orm import Session
 
@@ -34,10 +34,12 @@ if t.TYPE_CHECKING:
     from meltano.core.plugin.settings_service import PluginSettingsService
     from meltano.core.project import Project
 
-    PluginSettingsServiceFactory: t.TypeAlias = Callable[
-        [ProjectPlugin],
-        PluginSettingsService,
-    ]
+    class PluginSettingsServiceFactory(t.Protocol):
+        def __call__(
+            self,
+            plugin: ProjectPlugin,
+            env_override: dict | None = None,
+        ) -> PluginSettingsService: ...
 
 
 @pytest.mark.order(0)
@@ -86,7 +88,7 @@ def subject(tap, plugin_settings_service_factory) -> PluginSettingsService:
 
 
 @pytest.fixture
-def environment(project: Project) -> Generator[Environment, None, None]:
+def environment(project: Project) -> Generator[Environment | None, None, None]:
     project.activate_environment("dev")
     try:
         yield project.environment
@@ -223,7 +225,10 @@ class TestPluginSettingsService:
         """Casting is disabled for expandable strings."""
         monkeypatch.setenv("PORT", "4444")
         service = plugin_settings_service_factory(inherited_tap)
+
         parent = service.inherited_settings_service
+        assert parent is not None
+
         parent.set(
             "port",
             "5555",
@@ -1093,7 +1098,7 @@ class TestPluginSettingsService:
 
     def test_inherited_env_override(
         self,
-        plugin_settings_service_factory,
+        plugin_settings_service_factory: PluginSettingsServiceFactory,
         inherited_tap: ProjectPlugin,
         monkeypatch: pytest.MonkeyPatch,
     ) -> None:

--- a/tests/meltano/core/plugin/test_plugin_settings.py
+++ b/tests/meltano/core/plugin/test_plugin_settings.py
@@ -1103,7 +1103,7 @@ class TestPluginSettingsService:
         parent_service = plugin_settings_service_factory(tap)
         parent_env_key = "DBT_POSTGRES_HOST"
         parent_env_val = "parent-host"
-        
+
         # Manually inject into the parent's environment context
         monkeypatch.setitem(parent_service.env_override, parent_env_key, parent_env_val)
 

--- a/tests/meltano/core/plugin/test_plugin_settings.py
+++ b/tests/meltano/core/plugin/test_plugin_settings.py
@@ -1100,7 +1100,7 @@ class TestPluginSettingsService:
         # 1. Initialize the child service (which builds its parent)
         child_service = plugin_settings_service_factory(inherited_tap)
         parent_service = child_service.inherited_settings_service
-        
+
         # 2. Assert they are distinct memory objects (shallow copy)
         assert child_service.env_override is not parent_service.env_override
 

